### ISSUE TITLE
Feature Request: Supporting Proxy Protocol v2

### ISSUE THREAD

**Issue by alibo** [metallb/metallb#797]
Hi everyone!

### Problem

It's not possible to preserve source IP (client IP) when `service.spec.externalTrafficPolicy` is set to `Cluster`.

### Current Workaround

The current workaround is setting `service.spec.externalTrafficPolicy` to `Local`. But it has its own limitations:

1. The incoming traffic not balanced (when some pods are running on the same node)
2. IP of all worker nodes should be advertised with BGP and it increases the number of records of the routing table in the routers. Also, some routers have limitations for load-balancing the traffic between nodes more than X (16?) nodes.

#### What is "proxy protocol v2"?

https://developers.cloudflare.com/spectrum/proxy-protocol

#### Which LB providers have already implemented this feature?

- AWS: https://kubernetes.io/docs/concepts/services-networking/service/#proxy-protocol-support-on-aws (`service.beta.kubernetes.io/aws-load-balancer-proxy-protocol`)
- IBM: https://www.ibm.com/cloud/blog/know-your-real-clients-use-proxy-protocol-on-ibm-cloud-kubernetes-service (`service.kubernetes.io/ibm-load-balancer-cloud-provider-enable-features: "proxy-protocol"`)
- F5: https://devcentral.f5.com/s/articles/How-to-persist-real-IP-into-Kubernetes
- Digitalocean: https://www.digitalocean.com/blog/load-balancers-now-support-proxy-protocol/
- OVH: https://docs.ovh.com/gb/en/load-balancer/proxyprotocol/

### Implementation Ideas

Unfortunately, I'm not familiar with the code base of metallb, but I've investigated a little bit and I guess we can use `mmproxy` (from cloudflare) or `go-mmproxy` (from path.net) and deploy it as a sidecar to the `speaker` pods, but it's just a guess and I don't know the details. Also, there might be other options as well.

- https://blog.cloudflare.com/mmproxy-creative-way-of-preserving-client-ips-in-spectrum/
- https://blog.path.net/introducing-go-mmproxy/

**Comment by m-yosefpor** [comment]
+1 . That feature would be really nice to have.

**Comment by champtar** [comment]
Hi @alibo @m-yosefpor 

Another better way to preserve sourceIP is to use a CNI that support Direct Server Return, see #796.
MetalLB is controlplane only, making it part of the dataplane means we would be responsible for the performance of the system, so more bugs to fight, I personally don't see that happening.
Also Calico will soon be able to annonce LoadBalancers IP via BGP https://github.com/metallb/metallb/issues/114#issuecomment-754839513.

**Comment by russellb** [comment]
Based on @champtar 's comment, I think we can close this as he's clarified that MetalLB wouldn't implement this

**Comment by BloodyIron** [comment]
For those of us who do not want, nor have need, nor other justification, for BGP, things like Proxy Protocol (v1 & v2) support really is needed. I am currently trying to figure out a way to retain the true source-ip from traffic behind MetalLB Layer 2 (non BGP) and NGINX Ingress/others. I have looked into the Direct Server Return method and not really found any solutions, plus the also alternative presented of Calico presenting LoadBalancer IPs via BGP is not relevant, since.... BGP is not in-scope.

When I look at examples like this one ( https://docs.ovh.com/ca/en/kubernetes/getting-source-ip-behind-loadbalancer/ ) it looks like other load balancers can handle this, but Metal LB doesn't.

I can appreciate wanting to make decisions about how much additional work to take on, however the impact of not having solutions to retain source-ip for traffic means that in that mode (Layer 2 without BGP) an administrator cannot realistically track which external IP goes to which system, which from a security/abuse perspective makes that job effectively impossible.

Can we please re-open this issue and revisit the topic? I don't really see this particular functional need going away. So having Metal LB support some way to retain the source-ip in Layer 2 mode without BGP is something I for one (and I think others too) really want to see. Since I'm specifically going with Metal LB because Layer 2 without BGP is exactly what I need for this on-prem infrastructure.

Please?

**Comment by BloodyIron** [comment]
To be more explicit, the guide I linked fell on its face for me when I got to the part declaring "service.beta.kubernetes.io/ovh-loadbalancer-proxy-protocol" which clearly is an OVH-centric feature, not likely to work for me. Since... Metal LB does not implement Proxy Protocol (as per this thread).

Sadface.

**Comment by BloodyIron** [comment]
So if there's any way we can get this topic revisited @champtar / @russellb , I think it would be appreciated, please!

**Comment by alibo** [comment]
@BloodyIron have you tried Cilium with DSR-mode enabled? 
https://docs.cilium.io/en/v1.12/gettingstarted/kubeproxy-free/#direct-server-return-dsr

We've migrated to Cilium and we have enabled the `hybrid` loadbalancer mode in production. In this mode, the client IP is preserved for TCP traffic. (we could also enable it for UDP traffic, but because of packet fragmentation, we've disabled it)

**Comment by BloodyIron** [comment]
@alibo we're currently using Calico, can't remember this moment why I decided not to go with Cilium.

Are you using Metal LB in Layer 2 mode without BGP?

I appreciate you bringing this to my attention! I am not sure if this is a direction we want to head, as I will need to evaluate that, but yeah... I'll have to check :)

**Comment by alibo** [comment]
@BloodyIron we're using Metal LB in L3, but the incoming packets are SNATed by kubeproxy/cni itself. So I don't see any blocker for your requirement. Also, Calico supports this as it works similarly:
https://www.tigera.io/blog/hands-on-with-calicos-ebpf-service-handling/

**Comment by BloodyIron** [comment]
@alibo I'm reading through this, thanks! I'm uncertain though, can this be achieved without BGP? BGP is not in-scope nor in the environment, and this is on-prem in my case.

**Comment by alibo** [comment]
I'm glad it was helpful for you. 
I'm afraid, we haven't tested it, but in theory, as it uses eBPF for this, it should work.

**Comment by BloodyIron** [comment]
@alibo I can't seem to clearly tell if eBPF with Calico can work without BGP in the way I'm doing this. But I'm trucking through it anyways! Early on in planning for this cluster I decided Calico was the CNI I would go with (as I had to make a decision in planning), and I guess it may pay off. Original idea was "I'll use Calico, and come back to it when I want to use those features, it will already be there waiting for me", guess that was a good idea.

So far I have broken my cluster by just turning eBPF on, but was able to repair it by rebooting all nodes (NO THIS IS NOT ACTUAL PRODUCTION JUST YET).

I'm really hoping this can work without me having to restructure lots of stuff!

One other question.. do I _REALLY_ need to open up more ports on each of my node manually, or will Rancher/RKE1/Calico just handle that on my behalf? I'm not scared of more ports being open, but I'm trying to automate to the max here, and minimise manual steps when provisioning a node. What's your experience on this aspect?

Node's are Ubuntu 22.04 btw

**Comment by BloodyIron** [comment]
@alibo I'm not really making "visible progress" with Calico's eBPF, and I found this thread : https://github.com/metallb/metallb/pull/796

In that thread, when digging deeper into this specific comment ( https://github.com/metallb/metallb/pull/796#issuecomment-754200248 )

It seems that kube-router might be a good alternative, and is listed as working with MetalLB (the only problem is BGP and I don't care about BGP).

As for my Calico eBPF... it LOOKS like it's enabled, in that, the Felix parameter is accepted, but none of the calico-nodes mention if BPF is enabled/failed, or anything. No BPF mentioned in the logs at all, which doesn't match up with the Calico documentation on how to troubleshoot the current state of eBPF. Also, current testing shows that the Source IP does not appear to be maintained whether using DSR or Tunnel mode, so I'm not even getting the effect I want anyways.

I also want to add, for the history books, that since I've used Rancher to provision my k8s cluster/nodes, using RKE1, kube-proxy is present on each node, but I can't find it defined within the cluster, so it seems to be handled outside the k8s cluster in some way... And the Calico documentation (v3.24, but not v3.22 docs) mentions that RKE/Rancher is a scenario where Calico eBPF and Kube-Proxy can exist at the same time, but I cannot find what I need to do after "enabling eBPF" to get traffic through eBPF and seemingly not-kube-proxy.

So... I'm going to read more into this kube-router thing, as I think I may have exhausted all relevant resources for Calico's eBPF for the time being...

I suspect this would all be easier if Proxy Protocol v2 was just supported, but... yeah... my adventure continues... hours/days later :/

**Comment by BloodyIron** [comment]
Okay so I dug more into kube-router, and RKE1 looks to "support" it, but I can only find that from this : https://github.com/rancher/rke/issues/312

None of the release notes from releases around the time it was merged (or not too long after) mention kube-router, so it's probably in the version of RKE1 I'm using... but I can't explicitly confirm that.

Furthermore, it seems that there's no "nice" way to switch from Calico to kube-router, or even create a Cluster with kube-router, since Rancher (which I use to create the cluster, and re-create when validating DR) doesn't even present kube-router as an option, and I think that was overlooked due to this sole issue : https://github.com/rancher/rancher/issues/15470

So at this point. I don't think I have any solution for getting the "Source IP" working.

Again, this would be solved by having Proxy Protocol v2 support baked into Metal LB... like so many other LoadBalancers out there... And I still have no solution for retaining the true "Source IP" after many hours/days of toil on this subject.

I'm open to more ideas on what my options are, but... I am currently standing down and accepting at this time "Source IP" is unachievable for me... hopefully I discover something achievable and not insane as a solution.

**Comment by MohammedNoureldin** [comment]
Hi @BloodyIron!

Have you got any further? I am trying to avoid abandoning MetalLB and using another load balancing solution only because of this. I like how MetalLB works, but real source IP preserving through Proxy Protocol is just crucial.

I am interested in knowing if anyone found a reasonable solution using L3.

**Comment by alibo** [comment]
@MohammedNoureldin You can still use MetalLB but change your CNI to cilium with DSR enabled:
https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#dsr-mode

But, Cilium can also do what MetalLB does as well. (for BGP (L3) mode, not ARP (L2) mode: https://github.com/cilium/cilium/issues/22246 )

**Comment by alibo** [comment]
^ Seems Cilium also supports L2 announcement in the recent version (v1.14): https://docs.cilium.io/en/stable/network/l2-announcements/

**Comment by MohammedNoureldin** [comment]
@alibo as far as you know, isn't it possible somehow with calico?

**Comment by alibo** [comment]
@MohammedNoureldin we've tested Cilium, I'm afraid, I'm not sure how Calico works

**Comment by MohammedNoureldin** [comment]
@alibo thanks!

Though it does not seem for me as a proper solution to turn kube proxy totally off. I would keep using it as standard, but avoid it for some services, if at all possible. Still the easiest way is if MetalLB implements proxy protocol.

**Comment by alibo** [comment]
> Though it does not seem for me as a proper solution to turn kube proxy totally off. I would keep using it as standard

@MohammedNoureldin I'm afraid, with the current implementation, the real problem is the kube-proxy itself. when a service with the type Loadbalancer is created, K8s uses NodePort in the bottom layer to expose the service. Kube-proxy uses IPTable rules to SNATs / Masurades the packets to forward the incoming packets from a node to a pod running on another node. That's why the actual source IP is lost. 

You may want to check these files for more information:
- https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/README.md
- https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go
- https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go

On the other hand, Cilium uses eBPF to inject the required metadata (encoding actual source IP and port in the SYN packet for TCP packets and every UDP packet if the full DSR mode is enabled) into the packets to preserve the source IP after the traffic is NATed in the origin node. When it's passed to the destination node, the source IP of the packet (which is the origin node right now) will be replaced with the injected metadata (actual source IP and port) by the Cilium agent running on the destination node utilizing the eBPF functionality. 

![dsr-without](https://github.com/metallb/metallb/assets/5338482/1b80e89d-c72a-4d90-9f8a-978f3d0ffd87)
![dsr-with](https://github.com/metallb/metallb/assets/5338482/6e1e1a04-e4b3-4c17-a242-240cacfd1915)

More info: 
- https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#direct-server-return-dsr
- https://www.youtube.com/watch?v=zoxQPI91KWk

**Comment by BloodyIron** [comment]
I haven't gotten to the point of PoCing for results, but currently I'm working towards (in a testing k8s cluster) ripping out my CNI and replacing it with kube-router. Not yet sure in what mode, but in my... unreasonably exhaustive and avoidable research (I'm looking at you MetalLB Devs opting out of Proxy Protocol 2), it seems like kube-router is maybe the only (or best in my case?) solution for preserving the Source-IP data. Also, kube-router seems to be faster for performance, so yay to that!

Not sure how long it will take before I get my results, lots on my plate, but this could be a worthwhile avenue for others interested.

Oh and yes, this PoC intent is to nuke-from-orbit kube-proxy as much as I can (hopefully I can fully disable/remove it, but I may have to "tolerate" it running and doing nothing).

**Comment by BloodyIron** [comment]
Just an update. I have not ripped out my CNI, as I'm currently using Calico and exploring the eBPF capabilities with it. I am getting "libbpf: prog 'calico_connect_v4': failed to attach" when I try to enable it, and not yet sure what I need to do to solve it. There are barely any resources on the internet on the matter (half of which are my posts on the internet asking for help, lol). So it's not yet clear WTF I need to do to get it going... but I do have the most confidence in Calico+eBPF right now, as using kube-router instead of kube-proxy looks extremely hard to do with RKE1 (as I don't yet see a way to disable kube-proxy with RKE1 nodes, so I'm probably going to have that just still running while Calico does eBPF stuff whenever I figure that out).

Just adding notes here for other humans.

**Comment by BloodyIron** [comment]
I _STILL_ cannot find a way to preserve the Source IP with self-hosted. I tried switching to Cilium except that breaks itself in ways that I cannot find any reason why it breaks. I was able to get SourceIP with it but having reachability to my pods break after literally random periods of time is just completely non-workable.

I've also tried endless permutations of other methods. Not only completely unsuccessfully, but no actually reliable resources online for me to implement that fits all the details of the picture.

You know what might work? Is if MetalLB had a way to preserve the SourceIP information. Can we PLEASE get the mechanism to do that already? I have MetaLB reliably working _EXCEPT_ for the single facet of not having the means to provide the SourceIP. And that not being a priority for the devs really is just completely absurd. It's extremely valuable information to preserve.

**Comment by fedepaol** [comment]
MetalLB is only about advertising the LB ips. It's not part of the dataplane. Once the packet lands on the node, all the rest **including changing the source ip** is performed by the CNI.

I suggest redirecting your energy towards the CNI of your choice, as it's the right place to preserve the src ip.

**Comment by axgkl** [comment]
Hey. 

> MetalLB is only about advertising the LB ips.

Thing is: Now that I understand the k8s world and its abstractions a bit better it is all clear and makes a ton of sense.

It's all about communication, imho - as are most problems between us humans :-)

=> I _suggest_ to explain the "as much as possible" in here, in a sentence or two - and a lot of frustration might be avoided:

![2024-07-18_1332x562_scrot](https://github.com/user-attachments/assets/96c50b28-69d1-4668-ae23-5175e1223c4e)


<details><summary>for your amusement ... ;-)</summary>
 First of all: This is totally a skill issue - but maybe consider, that this here is a component, which also noobs in this domain absolutely need to have, to get *anything* up, presentable.
 
 
And: Yep,  source ip is a must have in this day and age, especially here in Europe, showstopper otherwise. 


And: Noobs like me are just misled by that first paragraph on the metal lb homepage.

It made me think, wow, 
- yes, exactly my problem (pending lb),
- yes, requirements all fulfilled - and 
- yes, exactly targeting my use case, not having a cloud provider lb available, in all situations where customers want on prem *and* k8s.

Maybe also the naming - misled me (and as it seems others as well) to think: yep here hetzner, ...whatever lb, with that proxyProtocol support as a checkbox, a feature switch - and here is metal lb, with the mission statement to erradicate the need of the former - and it seems to be *the* default solution out there. So one assumes (naively), it is also just a switch, to care about, when all the other stuff is solved, to get the system presentable and automatically up.

And then your realize - ups,  your legs are gone :-)

So yeah, would have avoided a bad day at work, that clarification.

</details>

No stress, just my 5 cents. Thanks for this great software!

**Comment by BloodyIron** [comment]
I'm currently working to test RKE2 nodes + Calico in eBPF mode with MetalLB still doing ARP Layer 2 for a single IP load-balancing. My previous Cilium attempts worked except Cilium was extremely unreliable for reasons that I could not find, and Cilium would break itself 100% of the time after random durations in the range of 5min-1h30min, and it was obnoxious to repair. No self-healing was working for it. And there were other Disaster Recovery scenarios where Cilium failed. Nothing in the logs were showing why, I could not find any online resources on solutions, so I'm considering Cilium unreliable (despite reliably retaining the correct SourceIP, argh).

I am optimistic about my upcoming RKE2 + Calico testing, as RKE2 is much easier to disable kube-proxy on. But I prefer birds in the hand, not in the bush.

**Comment by axgkl** [comment]
>  and Cilium would break itself 100% of the time after random durations in the range of 5min-1h30min

:sweat: We decided to go for Cilium, do you have details when this happens? Heavy load or what triggered it? Which routingMode and tunnelProtocol, if I may ask?
---
Regarding the source ip: We are using external caddy l4 proxy or hetzner lb, with proxy ip set and internally, in the cluster an nginx ingress which we configure like this


```yaml
controller:
  kind: DaemonSet
  service:
    externalIPs: []
    nodePorts:
      http: 30080
      https: 30443
  config:
    # https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/
    use-proxy-protocol: "true"
 ```
 (if hetzner lb than with ccm annotations)
 
 That way nginx reads the proxyProtocol header and inserts the ip nicely into the headers. That's all independent of the CNI in use, and I guess that's how its done typically - nothing to do with metallb.

**Comment by BloodyIron** [comment]
> > and Cilium would break itself 100% of the time after random durations in the range of 5min-1h30min
> 
> 😓 We decided to go for Cilium, do you have details when this happens? Heavy load or what triggered it? Which routingMode and tunnelProtocol, if I may ask?
> 
> Regarding the source ip: We are using external caddy l4 proxy or hetzner lb, with proxy ip set and internally, in the cluster an nginx ingress which we configure like this
> 
> ```yaml
> controller:
>   kind: DaemonSet
>   service:
>     externalIPs: []
>     nodePorts:
>       http: 30080
>       https: 30443
>   config:
>     # https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/
>     use-proxy-protocol: "true"
> ```
> 
> (if hetzner lb than with ccm annotations)
> 
> That way nginx reads the proxyProtocol header and inserts the ip nicely into the headers. That's all independent of the CNI in use, and I guess that's how its done typically - nothing to do with metallb.

For me it happens literally standing still. I would LOVE to be able to explain it, but for the life of me could not find any consistent patterns in any logs :(

This was in a dev/test k8s cluster simply to do PoC for SourceIP stuff. So an instance of NGINX Ingress (from k8s group), librespeed, and Cilium was generally the majority of noteworthy aspects. Otherwise pretty bog-standard basic self-hosted RKE2 provisioned by Rancher. 3x cluster nodes, each node doing all 3x functions (etcd/worker/forgetTheThirdName).

I had Cilium configured to do Layer 2 ARP LB as that is a functional requirement. In this case for a single IP for inbound.

The scope of the requirements is the LB is inside the cluster as we don't want yet-another-aspect that is manually handled outside the cluster. This is about migrating as much "stuff" as we can from VM->k8s+IaC-stuff.

There's no Proxy Protocol stuff active that I know of.

**Comment by cunningr** [comment]
Not sure if it was mentioned in already in this thread but for those interested in a MetalLB solution with proxy-protocol-v2, it might be possible to combine an ingress controller (such as [HAProxy](https://www.haproxy.com/documentation/kubernetes-ingress/overview/)) with MetalLB and do to TCP load-balancing.  If you couple HAProxy with a LoadBalancer in `Local` model, you will get the src IP.  HAProxy controller can then be configured using k8s manifests to load share tcp and (and/or http) to workloads behind an internal k8s service.  This kind of adds a dataplane to MetalLB.
